File: bindings/swift/Sources/Proven/SafeUUID.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// UUID operations delegated to libproven FFI.
///
/// RFC 4122 UUID generation, parsing, and formatting via the
/// formally verified Idris 2 core.

import CProven

/// A UUID value (128-bit identifier).
public struct ProvenUUIDValue: Equatable, Sendable {
    /// The 16 bytes of the UUID.
    public let bytes: (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8,
                       UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
}

public enum SafeUUID {
    /// Generate a random v4 UUID.
    public static func v4() -> Result<ProvenUUIDValue, ProvenError> {
        let result = proven_uuid_v4()
        if let error = ProvenError.fromStatus(result.status) {
            return .failure(error)
        }
        let b = result.uuid.bytes
        return .success(ProvenUUIDValue(bytes: (
            b.0, b.1, b.2, b.3, b.4, b.5, b.6, b.7,
            b.8, b.9, b.10, b.11, b.12, b.13, b.14, b.15
        )))
    }

    /// Format a UUID as canonical string (8-4-4-4-12).
    public static func toString(_ uuid: ProvenUUIDValue) -> Result<String, ProvenError> {
        var cUuid = ProvenUUID()
        let b = uuid.bytes
        cUuid.bytes.0 = b.0; cUuid.bytes.1 = b.1; cUuid.bytes.2 = b.2; cUuid.bytes.3 = b.3
        cUuid.bytes.4 = b.4; cUuid.bytes.5 = b.5; cUuid.bytes.6 = b.6; cUuid.bytes.7 = b.7
        cUuid.bytes.8 = b.8; cUuid.bytes.9 = b.9; cUuid.bytes.10 = b.10; cUuid.bytes.11 = b.11
        cUuid.bytes.12 = b.12; cUuid.bytes.13 = b.13; cUuid.bytes.14 = b.14; cUuid.bytes.15 = b.15
        return consumeStringResult(proven_uuid_to_string(cUuid))
    }

    /// Parse a UUID from canonical string format.
    public static func parse(_ string: String) -> Result<ProvenUUIDValue, ProvenError> {
        withStringBytes(string) { ptr, len in
            let result = proven_uuid_parse(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }
            let b = result.uuid.bytes
            return .success(ProvenUUIDValue(bytes: (
                b.0, b.1, b.2, b.3, b.4, b.5, b.6, b.7,
                b.8, b.9, b.10, b.11, b.12, b.13, b.14, b.15
            )))
        }
    }

    /// Check if a UUID is the nil UUID (all zeros).
    public static func isNil(_ uuid: ProvenUUIDValue) -> Bool {
        var cUuid = ProvenUUID()
        let b = uuid.bytes
        cUuid.bytes.0 = b.0; cUuid.bytes.1 = b.1; cUuid.bytes.2 = b.2; cUuid.bytes.3 = b.3
        cUuid.bytes.4 = b.4; cUuid.bytes.5 = b.5; cUuid.bytes.6 = b.6; cUuid.bytes.7 = b.7
        cUuid.bytes.8 = b.8; cUuid.bytes.9 = b.9; cUuid.bytes.10 = b.10; cUuid.bytes.11 = b.11
        cUuid.bytes.12 = b.12; cUuid.bytes.13 = b.13; cUuid.bytes.14 = b.14; cUuid.bytes.15 = b.15
        return proven_uuid_is_nil(cUuid)
    }

    /// Get the UUID version number.
    public static func version(_ uuid: ProvenUUIDValue) -> UInt8 {
        var cUuid = ProvenUUID()
        let b = uuid.bytes
        cUuid.bytes.0 = b.0; cUuid.bytes.1 = b.1; cUuid.bytes.2 = b.2; cUuid.bytes.3 = b.3
        cUuid.bytes.4 = b.4; cUuid.bytes.5 = b.5; cUuid.bytes.6 = b.6; cUuid.bytes.7 = b.7
        cUuid.bytes.8 = b.8; cUuid.bytes.9 = b.9; cUuid.bytes.10 = b.10; cUuid.bytes.11 = b.11
        cUuid.bytes.12 = b.12; cUuid.bytes.13 = b.13; cUuid.bytes.14 = b.14; cUuid.bytes.15 = b.15
        return proven_uuid_version(cUuid)
    }
}

File: bindings/fortran/src/safe_uuid.f90
! SPDX-License-Identifier: PMPL-1.0-or-later
! Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
!
! Proven SafeUUID - UUID generation and validation for Fortran
!
! Thin wrapper around libproven's verified SafeUUID module.
! All UUID logic (generation, parsing, formatting) is executed in the
! Idris 2 core via the Zig FFI bridge; this module only marshals types
! between Fortran and C.
!
! Link with: -lproven

module safe_uuid
    use, intrinsic :: iso_c_binding
    use proven_ffi
    implicit none
    private

    !> UUID type (Fortran-facing)
    type, public :: uuid_type
        integer(c_int8_t) :: bytes(16) = 0_c_int8_t
        logical            :: valid     = .false.
    end type uuid_type

    !> UUID parse/generate result type
    type, public :: UuidResult
        type(uuid_type)    :: uuid
        logical            :: ok     = .false.
        integer(c_int32_t) :: status = PROVEN_ERR_NOT_IMPLEMENTED
    end type UuidResult

    public :: generate_uuid_v4, parse_uuid, format_uuid
    public :: uuid_is_nil, uuid_get_version

contains

    ! =========================================================================
    ! Generate UUID v4  ->  proven_uuid_v4
    ! =========================================================================
    function generate_uuid_v4() result(res)
        type(UuidResult) :: res
        type(c_proven_uuid_result) :: c_res

        c_res = proven_uuid_v4()

        res%status    = c_res%status
        res%ok        = (c_res%status == PROVEN_OK)
        res%uuid%bytes = c_res%uuid%bytes
        res%uuid%valid = res%ok
    end function generate_uuid_v4

    ! =========================================================================
    ! Parse UUID from string  ->  proven_uuid_parse
    ! =========================================================================
    function parse_uuid(uuid_string) result(res)
        character(len=*), intent(in) :: uuid_string
        type(UuidResult) :: res
        character(len=len_trim(uuid_string), kind=c_char), target :: c_buf
        type(c_proven_uuid_result) :: c_res
        integer :: trimmed_len

        trimmed_len = len_trim(uuid_string)
        c_buf = uuid_string(1:trimmed_len)

        c_res = proven_uuid_parse(c_loc(c_buf), int(trimmed_len, c_size_t))

        res%status    = c_res%status
        res%ok        = (c_res%status == PROVEN_OK)
        res%uuid%bytes = c_res%uuid%bytes
        res%uuid%valid = res%ok
    end function parse_uuid

    ! =========================================================================
    ! Format UUID as string  ->  proven_uuid_to_string
    ! =========================================================================
    function format_uuid(uuid_instance) result(uuid_string)
        type(uuid_type), intent(in) :: uuid_instance
        character(len=36) :: uuid_string
        type(c_proven_uuid) :: c_uuid
        type(c_proven_string_result) :: c_res
        character(len=1, kind=c_char), pointer :: f_ptr(:)

        uuid_string = ''
        c_uuid%bytes = uuid_instance%bytes

        c_res = proven_uuid_to_string(c_uuid)

        if (c_res%status == PROVEN_OK .and. c_associated(c_res%value)) then
            call c_f_pointer(c_res%value, f_ptr, [c_res%length])
            block
                integer :: i, copy_len
                copy_len = min(int(c_res%length), 36)
                do i = 1, copy_len
                    uuid_string(i:i) = f_ptr(i)
                end do
            end block
            call proven_free_string(c_res%value)
        end if
    end function format_uuid

    ! =========================================================================
    ! Check if UUID is nil  ->  proven_uuid_is_nil
    ! =========================================================================
    function uuid_is_nil(uuid_instance) result(is_nil)
        type(uuid_type), intent(in) :: uuid_instance
        logical :: is_nil
        type(c_proven_uuid) :: c_uuid

        c_uuid%bytes = uuid_instance%bytes
        is_nil = logical(proven_uuid_is_nil(c_uuid))
    end function uuid_is_nil

    ! =========================================================================
    ! Get UUID version  ->  proven_uuid_version
    ! =========================================================================
    function uuid_get_version(uuid_instance) result(ver)
        type(uuid_type), intent(in) :: uuid_instance
        integer :: ver
        type(c_proven_uuid) :: c_uuid

        c_uuid%bytes = uuid_instance%bytes
        ver = int(proven_uuid_version(c_uuid))
    end function uuid_get_version

end module safe_uuid
